src/liballoc/string.rs

```diff
@@ -1225,6 +1225,62 @@ impl String {
         ch
     }
 
+    /// Remove all matches of pattern `pat` in the `String`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(string_remove_matches)]
+    /// let mut s = String::from("Trees are not green, the sky is not blue.");
+    /// s.remove_matches("not ");
+    /// assert_eq!("Trees are green, the sky is blue.", s);
+    /// ```
+    ///
+    /// Matches will be detected and removed iteratively, so in cases where
+    /// patterns overlap, only the first pattern will be removed:
+    ///
+    /// ```
+    /// #![feature(string_remove_matches)]
+    /// let mut s = String::from("banana");
+    /// s.remove_matches("ana");
+    /// assert_eq!("bna", s);
+    /// ```
+    #[unstable(feature = "string_remove_matches", reason = "new API", issue = "none")]
+    pub fn remove_matches<'a, P>(&'a mut self, pat: P)
+    where
+        P: for<'x> Pattern<'x>,
```

Comment:
This bound is pretty unusual. Does something block this from being implementable with only a `P: Pattern<'a>` bound?

Reply:
It wouldn't compile without the higher-rank trait bound. I talked to someone on the community Discord about this and they suggested the HRTB and it made it compile.

I think it had something to do with not having the HRTB meant Rust thought `pat` was going to live as long as `self` was, when in fact the lifetime of `pat` had to be shorter.

Reply:
If I change `P: for<'x> Pattern<'x>` to `P: Patten<'a>` compiling fails with:
```rust
error[E0502]: cannot borrow `self.vec` as mutable because it is also borrowed as immutable
    --> src/liballoc/string.rs:1274:21
     |
1249 |     pub fn remove_matches<'a, P>(&'a mut self, pat: P)
     |                           -- lifetime `'a` defined here
...
1256 |             let mut searcher = pat.into_searcher(self);
     |                                -----------------------
     |                                |                 |
     |                                |                 immutable borrow occurs here
     |                                argument requires that `*self` is borrowed for `'a`
...
1274 |                     self.vec.as_mut_ptr().add(end - shrunk_by),
     |                     ^^^^^^^^ mutable borrow occurs here

error[E0502]: cannot borrow `self.vec` as mutable because it is also borrowed as immutable
    --> src/liballoc/string.rs:1275:21
     |
1249 |     pub fn remove_matches<'a, P>(&'a mut self, pat: P)
     |                           -- lifetime `'a` defined here
...
1256 |             let mut searcher = pat.into_searcher(self);
     |                                -----------------------
     |                                |                 |
     |                                |                 immutable borrow occurs here
     |                                argument requires that `*self` is borrowed for `'a`
...
1275 |                     self.vec.as_mut_ptr().add(start - shrunk_by),
     |                     ^^^^^^^^ mutable borrow occurs here

error[E0502]: cannot borrow `self.vec` as mutable because it is also borrowed as immutable
    --> src/liballoc/string.rs:1280:13
     |
1249 |     pub fn remove_matches<'a, P>(&'a mut self, pat: P)
     |                           -- lifetime `'a` defined here
...
1256 |             let mut searcher = pat.into_searcher(self);
     |                                -----------------------
     |                                |                 |
     |                                |                 immutable borrow occurs here
     |                                argument requires that `*self` is borrowed for `'a`
...
1280 |             self.vec.set_len(len - shrunk_by);
     |             ^^^^^^^^ mutable borrow occurs here

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0502`.
error: could not compile `alloc`.
```

Reply:
@dtolnay does this answer your question about why I used this trait bound?

Reply:
It does not -- it just says that the current *implementation* relies on the too strict bound. In https://github.com/rust-lang/rust/pull/71780#pullrequestreview-420582678 I wrote that we'd want to fix the implementation so that a less strict bound is used.

Reply:
For example the following code also requires an unnecessarily strict bound, and does not compile if you remove `'a: 'static`, but it's easy to fix the implementation to not require that bound.

```rust
pub fn first<'a, T>(input: &'a [T]) -> Option<&'a T>
where
    'a: 'static,
{
    let slice: &'static [T] = input;
    slice.get(0)
}
```